Support displaying bookmark state and watched progress in video manager

hambre opened the pull request:
This adds support of displaying the bookmark state and the watched progress
within the video manager.
I have added preliminary support to display bookmark state and watched progress to the Monochrome theme
as seen in the image below:
![Bookmark_and_WatchedProgress_in_VideoManager_annotated](https://user-images.githubusercontent.com/6901566/229840896-3d614fd9-d28d-4b16-90bb-1c5205b2209a.png)

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] contribution does not duplicate one of our [existing pull requests](https://github.com/MythTV/mythtv/pulls)
- [x] contribution is in a branch rebased against [master](https://github.com/MythTV/mythtv)
- [x] code compiles successfully without errors
- [x] code follows the [MythTV Coding Standards](https://www.mythtv.org/wiki/Coding_Standards)
- [x] documentation added/updated/removed where necessary
- [x] commits are logically organised and have [good commit messages](https://chris.beams.io/posts/git-commit)

rcrdnalor commented:
The terminus 'Plugin' has a special semantic within MythTV.
See 
https://www.mythtv.org/wiki/User_Manual:Plugins_and_Addons
https://www.mythtv.org/wiki/Category:Plugins
I see a 'MythMediaPlugin' in your proposed changeset.
IMHO, this change will confuse many of the developer and user.
If it is not a 'plugin' in the sense of MythTV, please consider a different naming.

hambre commented:
> The terminus 'Plugin' has a special semantic within MythTV. See https://www.mythtv.org/wiki/User_Manual:Plugins_and_Addons https://www.mythtv.org/wiki/Category:Plugins I see a 'MythMediaPlugin' in your proposed changeset. IMHO, this change will confuse many of the developer and user. If it is not a 'plugin' in the sense of MythTV, please consider a different naming.

Yes, no problem. I just followed the naming of the function to register the formerly used MediaCallback, which is called 'RegisterMediaPlugin'. I could rename 'MythMediaPlugin' to 'MythMediaHandler', but then the register function should be also renamed to avoid confusion.

warpme commented:
hambre
Nice work!

One Q: is this code allows to have also watched progressbars on video lists entries?
(the same like in tv recordings)?

hambre commented:
> The terminus 'Plugin' has a special semantic within MythTV. See https://www.mythtv.org/wiki/User_Manual:Plugins_and_Addons https://www.mythtv.org/wiki/Category:Plugins I see a 'MythMediaPlugin' in your proposed changeset. IMHO, this change will confuse many of the developer and user. If it is not a 'plugin' in the sense of MythTV, please consider a different naming.

hambre commented:
> hambre Nice work!
> 
> One Q: is this code allows to have also watched progressbars on video lists entries? (the same like in tv recordings)? I'm asking as i'm not seeing progress bars on my recordings. I applied [hambre/Monochrome@d485478](https://github.com/hambre/Monochrome/commit/d485478a5b37663847037205005cf8e508402796) and i' not sure what else i'm missing ?

Hi Warpme, yes it allows progress bars for video list entries. If you look at the left red mark on the screenshot, you can see a green progressbar under the video title. However for that to work the video length entry within the videomarkup table needs to be present.

warpme commented:
Ah - i missed green bar on screenshot :-)

Also - isn't be good idea also to add new field in filemarkup with total frames for movie?
With this - progress will be simple mark/total frames.
Maybe this can be fallback path - when user not run yet mythcommflag to rebuild seektable for a video file? 
Benefit is that showing progress feature will be available also without any extra user actions (like rebuild seektable).

hambre commented:
I think the total duration mark is given in frames. The problem is where to get this value from without relying on mythcommflag. I would love to add that, so I would not need a script to actually create this using mythcommflag. Any ideas?

warpme commented:
Well - simple idea might be to do (or estimate) total_frames in mythplayer.
When user watching video and press for osd to see progress - there is progress shown on osd - so we have her required info.
So for example we can use this i.e. to estimate total_frames to store it filemarkup table ?

warpme commented:
btw: i love your work with Monochrome!
imho: most beautiful theme with perfect form-beauty balance!

hambre commented:
Hmm ok, I can have a look to come up with something. Thanks for the hint!

hambre commented:
@warpme Thanks to your hint I added saving the total frame count for video files automatically without needing to use external tooling. You might want to check if that works for you.

warpme commented:
@hambre 
https://github.com/MythTV/mythtv/commit/2f86ddbf2ea70c6f4b723cc66a497cd1b7d3d58d works perfectly :-)

One Q: isn't worth to have behavior of video progress as much as possible similar to recordings progress behavior?
I mean: user see progress bar on recording when: status is "not watched" & progress is > 0.

Also - i see small issue when user enters videolist: there is no progress-bar on list entries unless user gives active focus on list entry. 
(This is different compared to progress bar behavior in recordings: entering recordings already shows progress bars on all recordings with status "not watched" & progress > 0)

hambre commented:
@warpme 
For me that is a little different. I typically watch recorded programs once. If I want to watch them again later I move them to the movies. So for movies which might be watched multiple times, I think having a progress bar not just for the first viewing is a feature I want to have.
For the progress bar on all items of the list, there is more work to do. That would need pulling all of the watched progress for all movies at once and caching that data to avoid too many queries to the database. So I did not do that first. I might look into that if time permits.

warpme commented:
@hambre 

You mention: 

> For me that is a little different. I typically watch recorded programs once. If I want to watch them again later I move them to the movies. So for movies which might be watched multiple times, I think having a progress bar not just for the first viewing is a feature I want to have.

I'm reading this as alternative use-case to current mythtv offers:

**mythv:**
"watched" means "is done" - so no progress make sense for something which is done. By this - when "watched" is ON, progress disappears (as not make sense anymore)

**your needs:**
Even is movie was already watched - you want to see progress of your's last watching - despite first watching is done 
 
So ideally will be to add setting like: "_always show watching progress_" which functionally will be opposite to "_show watching progress only for unwatched_" (like in current mythtv). 
Such setting should be mythtv wide (i mean it should work the same for recordings and for movies) 

With this we can cover: current presentation model & your alternative preferred model.... 

What you think?

hambre commented:
@warpme 

Yes this might work to support the different use cases. I will look into it.

hambre commented:
@warpme 
I reworked this PR completely to have the changes more contained and to support displaying bookmark and watched progress for all visible items. Also added a setting to enable displaying the watched progress even when a video/recording is already marked as watched (default: off).

warpme commented:
@hambre 
Works perfectly!
Really nice work. 
I hope it will be merged asap :-)

hambre commented:
It seems not. String freeze for Myth 34 is already in effect.

kmdewaal commented:
> It seems not. String freeze for Myth 34 is already in effect.

Can still be added to v34 as far as I know. It will be then only in English and not translated.

hambre commented:
> > It seems not. String freeze for Myth 34 is already in effect.
> 
> Can still be added to v34 as far as I know. It will be then only in English and not translated.

Good to know. It still needs one of the developers to look at the PR and hopefully merge it.

kmdewaal reviewed (approved):
Looks OK to me. It does not appear to break anything. To make it useful there must be changes in the themes but I expect that then to be added later. There is one translatable string added but I expect at least the German translation can still be done before v34 is really frozen.
I am OK with merging this now.

hambre commented:
> Looks OK to me. It does not appear to break anything. To make it useful there must be changes in the themes but I expect that then to be added later. There is one translatable string added but I expect at least the German translation can still be done before v34 is really frozen. I am OK with merging this now.

Thanks for your approval! I can do the German translation for the new string as soon as this is merged, no problem. However I do not have the permissions to merge myself, so you or one of the other devs need to do that.

hambre commented:
@kmdewaal Thanks for merging!
